rfcs/0031-devices-api.md

```diff
@@ -0,0 +1,363 @@
+
+```c
+#include <device.h>
+#include <kernel.h>
+
+typedef struct {
+    struct device* dev;
+    k_mutex lock;
+} tvm_device_woofles_t;
+```
+
+## Generic Device API
+The majority of the device API calls should be added to the platform-agnostic `<device>.h`:
+```c
+int32_t TVMDeviceWooflesActivate(tvm_device_woofles_t* tvm_dev); // Called by generated code
+int32_t TVMDeviceWooflesOpen(tvm_device_woofles_t* tvm_dev); // Called by generated code
+int32_t TVMDeviceWooflesClose(tvm_device_woofles_t* tvm_dev); // Called by generated code
+int32_t TVMDeviceWooflesDeactivate(tvm_device_woofles_t* tvm_dev); // Called by generated code
+```
+
+These can all be implemented using the user-opaque context `tvm_device_<device>_t*`, enabling the majority of TVM code to be portable between RTOS implementations; importantly this applies to those called within operator functions (see below). The executors are agnostic to the underlying device implementation and simply get passed the relevant device pointer which is then passed to the correct symbol.
+
+## Platform Device API
+To allow setting of platform specifics into the opaque struct, these should be defined in the platform header. Alongside the header, an additional file will provide implementations (`src/runtime/crt/device/<device>/<platform>.c`):
+```c
+int32_t TVMDeviceWooflesInit(tvm_device_t* tvm_dev, struct device* zephyr_dev) {
+    tvm_dev->device = zephyr_dev;
+}
+```
+This simple wrapper enables type checking of these functions and defining a clear translation boundary between the underlying OS implementation and TVM.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Entrypoint
+The entrypoint API defined in [Embedded C Runtime Interface](https://discuss.tvm.apache.org/t/rfc-utvm-embedded-c-runtime-interface/9951) is augmented with the `devices` structure which contains implemented `tvm_device_t` `struct`s for each device used by the network:
+```
+typedef struct {
+    struct tvm_device_woofles_t* woofles
+} tvmgen_mynetwork_devices;
+```
+
+These are re-cast to `void *` when entering the AOT main function to pass it without TIR understanding the struct types.
+
+```c
+int32_t tvmgen_mynetwork_run(
+    ...,
+    struct tvmgen_mynetwork_devices* devices
+) {
+    tvmgen_mynetwork_run_model(
+        ...,
+        devices->host,
+        devices->accelerator
+    );
+}
+```
+
+## Executor Function
+Each operator is provided with a single device object which can be abstracted and passed as the `void* resource_handle`. The main function calls into the device API to setup and teardown resources before and after each operator call.
+
+```c
+int32_t tvmgen_mynetwork_run_model(..., device0, device1) {
+    TVMDeviceWooflesActivate(device0); // Could reserve or enable certain circuitry ahead of time
```

Comment:
would be great to say something about the memory and perhaps include a note that memory copies into the device API-managed memory may occur before TVMDeviceWooflesOpen

Reply:
I added some words to that effect, essentially stating that the memory flow is outside this RFC as it stands right now but we should be aware that means it could happen at any time.